### DISCUSSION TITLE
ImageNode transparency

### DISCUSSION THREAD

**Discussion by osoikatatsumuri** [sceneview/sceneview#485] [Q&A] [answered]
Is there any way to change the transparency of an ImageNode?

**Comment by ThomasGorisse** [comment]
Good question !
The fragment shader should handle it since its blending mode is `transparent` and since it doesn't filter the alpha channel for the `baseColor` but I'm not sure if the `Sampler2D` can handle `Bitmap` transparency
Could you ask on the Filament repo if it's possible?
Maybe we will have to handle it the same way as the video texture, using a chroma key color.
A hot fix would be to use the `ViewNode` with an `ImageView` but it will be a lot less efficient (please don't do that!!! 😀)

**Reply by osoikatatsumuri** [reply]
Will try my best to figure it out, thanks!

**Reply by ThomasGorisse** [reply]
Please post your results here.

**Reply by osoikatatsumuri** [reply]
It works very well, after all. I create this function to change bitmap opacity to what i need. @ThomasGorisse 

```
private fun adjustOpacity(bitmap: Bitmap, alpha: Int): Bitmap? {
    val rgbImage = bitmap.copy(Bitmap.Config.ARGB_8888, true)

    if (alpha < 0 || alpha > 255) {
        return null
    }

    for (y in 0 until rgbImage.height) {
        for (x in 0 until rgbImage.width) {
            val argbPixel = rgbImage.getPixel(x, y)
            val newPixel = (argbPixel and 0x00FFFFFF) or (alpha shl 24)
            rgbImage.setPixel(x, y, newPixel)
        }
    }

    return rgbImage
}
```

And then after imageNode creation:

`
    imageNode.bitmap = adjustOpacity(imageNode.bitmap, desiredAlpha)
`

And result:
![IMAGE 2024-05-16 17:18:23](https://github.com/SceneView/sceneview-android/assets/85958109/ec22f729-91b9-4848-80c6-1f654837d763)

**Reply by osoikatatsumuri** [reply]
Ok, it works very strange. I use ARSceneView and it's opacity depends on distance from the camera to the plane.  The closer the image, the lower the alpha channel

**Reply by osoikatatsumuri** [reply, marked as answer]
@ThomasGorisse Yeah, that's works now! Even without modifications, or maybe back then i did something wrong!

```
private fun adjustOpacity(bitmap: Bitmap, alpha: Int): Bitmap? {
    val rgbImage = bitmap.copy(Bitmap.Config.ARGB_8888, true)

    if (alpha < 0 || alpha > 255) {
        return null
    }

    for (y in 0 until rgbImage.height) {
        for (x in 0 until rgbImage.width) {
            val argbPixel = rgbImage.getPixel(x, y)
            val newPixel = (argbPixel and 0x00FFFFFF) or (alpha shl 24)
            rgbImage.setPixel(x, y, newPixel)
        }
    }

    return rgbImage
}
```

`imageNode.bitmap  = adjustOpacity(imageNode.bitmap, desiredAlpha)`

upd by 19.05.2025: You may notice that when you reduce the transparency and reduce the image size, the transparency may come back. To avoid this, just switch from TextureSampler2D to TextureSamplerExternal while ImageNode initialization
